aspect: add --video-recenter

guidocella opened the pull request:
This resets ``--video-align-x`` and ``--video-align-y`` to 0 when the video becomes smaller than the OSD in the respective direction, e.g. by zooming out.

Unlike doing this by observing osd-dimensions in a script, this is done before rerendering, so you don't see the image being rendered not centered for an instant after zooming out before being rerendered centered.

Defaults to yes.

guidocella commented on video/out/vo.c:
```diff
@@ -230,12 +232,35 @@ static void read_opts(struct vo *vo)
+    if (!vo->params || !vo->opts->recenter)
+        return;
+
+    struct MPContext *mpctx = (struct MPContext *)vo->extra.wakeup_ctx;
```
How can se signal the property notifications?

kasper93 replied:
It shouldn't

guidocella replied:
Shouldn't what?

guidocella replied:
Thanks, done.

kasper93 commented:
I understand what it does, but could you also explain why it does it? We can see what it does in the code, but commit description should describe why it exists and why it is default.

guidocella commented:
Added an explanation of the purpose, also replaced OSD references with window, since OSD refers more to the text.

It should be noted that if you manually start with --video-align-x=-1 it doesn't always start centered, but `recenter()` at the end of `vo_create` doesn't work.

kasper93 commented:
> After zooming in, panning and zooming out, this is useful to recenter the video in the window.

I still don't understand why this is useful. I have `video-align-x=-1` to anchor video to left border of the screen. I zoom out. Why it resets to the center? Maybe it is useful for some uses, but I don't see why this should be default behavior of mpv?

guidocella commented:
Of course, it is not useful if you actually want to anchor to a border. If you start centered, then zoom in, pan around and reset the zoom, you normally want the video to return to the center like before, which looks better, and is how image viewers behave. I will move the default of yes to the image profile since it's mostly useful for that, I can't even figure out how to zoom and pan in VLC to see how it behaves.

kasper93 commented:
> Also it is currently not possible to use video-align- in zommed out state, because any update to vo options resets it to zero.

Is this expected?

guidocella commented:
Yeah it should stay centered even if you move it by accident, that's how image viewers behave, or at least the ones I'm trying. Also zoomed out is a simplification, there are many more options that affect whether the video is bigger than the window, video-aspect-override, video-unscaled, video-crop, video-scale, video-margin-ratio, it would not be a sane simple zoom check to predict whether alignment can be changed.

kasper93 commented:
So you basically want to ignore `video-align-` in some conditions? Why does it reset the user's provided value? Why is it in the vo.c instead of aspect.c? Why are `video-pan-` options not affected?

kasper93 commented on options/options.c:
```diff
@@ -250,6 +251,7 @@ const struct m_sub_options vo_sub_opts = {
         .scale_x = 1.0f,
         .scale_y = 1.0f,
         .auto_window_resize = true,
+        .recenter = false,
```
not needed

kasper93 commented on DOCS/man/options.rst:
```diff
@@ -1652,17 +1652,34 @@ Video
+
     This option is disabled if ``--keepaspect=no`` is used.
 
+``--video-recenter=<yes|no>``
```
```suggestion
``--video-align-center=<yes|no>``
```

I don't have good name for this, without making it too long, but `recenter` suggest is it a "command" and not an option. It should be a state.

guidocella replied:
Plenty of options are verbs though? align-center just sounds wrong.

kasper93 replied:
I don't understand what are we REcentering. If I zoom out, I keep the video center, not recenter, which implies previous uncentered state.

guidocella replied:
Well recentering really refers to displayed video rectangle and not to the video itself, same as video-pan and video-align, they move the output rectangle especially when zoomed in, video is just the shared prefix of these options. So if the video rectangle was aligned to the left, it gets realigned to the center after zooming out. But I can see how it can be confusing so we can go with `--video-keep-centered` if you prefer?

kasper93 commented on DOCS/man/options.rst:
```diff
@@ -1652,17 +1652,34 @@ Video
     If video and screen aspect match perfectly, these options do nothing.
 
+    Unlike ``--video-pan-x`` and ``--video-pan-y``, these don't go beyond the
+    video's boundaries or make the displayed rectangle drift off after zooming.
```
video's boundaries or window?

guidocella commented on video/out/aspect.c:
```diff
@@ -76,12 +76,15 @@ static void clamp_size(int size, int *start, int *end)
 {
     scaled_src_size *= powf(2, zoom) * scale;
     scaled_src_size = MPMAX(scaled_src_size, 1);
+    if (recenter && dst_size > scaled_src_size)
```
What is the difference? video-align does nothing when the image is the same size as the window.

kasper93 replied:
It makes the condition follow documentation. We want to center align when zoomed out. functionally it is no different, but it makes the intent clear.

guidocella replied:
Zoomed out in the documentation doesn't mean the zoom is reset, it means it decreases, and recentering already takes place with a positive zoom.

Say an image is 2000 pixels wide on a 1920 screen at video-zoom=3 and align-x is 1. If you zoom out to video-zoom=2, recentering already takes place and the image is rendered as if align-x was 0.

guidocella replied:
Made the docs clearer.

kasper93 replied:
I understand what it doesn't, but logically it make sense to reset it to zero when `dst_size == scaled_src_size` even if it already would be multiplied by zero below, it just makes it explicit.
